### PR TITLE
Implement EIP3074: Authcall

### PR DESCRIPTION
**Pull request by jochem-brouwer**
Implements https://eips.ethereum.org/EIPS/eip-3074

### CODE REVIEW

File: packages/vm/tests/api/EIPs/eip-3074-authcall.spec.ts [hunk 1]
```diff
@@ -0,0 +1,597 @@
...
+      const expected = preGas - preGas / 64n - 2n
+      st.equals(gasBigInt, expected, 'forwarded max call gas')
+
+      const expectedBalance = PREBALANCE - result.amountSpent - value
+      const account = await vm.stateManager.getAccount(callerAddress)
+
+      st.equals(account.balance, expectedBalance, 'caller balance ok')
+
+      const contractAccount = await vm.stateManager.getAccount(contractAddress)
+      st.equals(contractAccount.balance, 2n, 'contract balance ok')
```
Comment (jochem-brouwer) [inline packages/vm/tests/api/EIPs/eip-3074-authcall.spec.ts:448]: Extremely weird, this test fails :thinking:

File: packages/vm/tests/api/EIPs/eip-3074-authcall.spec.ts [hunk 2]
```diff
@@ -0,0 +1,597 @@
...
+      gasLimit: 1000000,
+      gasPrice: 10,
+    }).sign(callerPrivateKey)
+
+    const result = await vm.runTx({ tx, block })
+    st.ok(
+      result.execResult.exceptionError?.error === ERROR.AUTHCALL_UNSET,
+      'threw with right error'
+    )
+    st.ok(result.amountSpent === tx.gasPrice * tx.gasLimit, 'spent all gas')
```
Comment (ryanio) [inline packages/vm/tests/api/EIPs/eip-3074-authcall.spec.ts:475]: can use st.equal here
```suggestion
    st.equal(result.amountSpent, tx.gasPrice * tx.gasLimit, 'spent all gas')
```

File: packages/vm/src/evm/opcodes/functions.ts [hunk 3]
```diff
@@ -976,7 +981,71 @@ export const handlers: Map<number, OpHandler> = new Map([
...
+        retLength,
+      ] = runState.stack.popN(8)
+
+      const toAddress = new Address(addressToBuffer(addr))
+
+      const gasLimit = runState.messageGasLimit!
+      runState.messageGasLimit = undefined
+
+      let data = Buffer.alloc(0)
+      if (!(argsLength === BigInt(0))) {
```
Comment (ryanio) [inline packages/vm/src/evm/opcodes/functions.ts:1038]: nit
```suggestion
      if (argsLength !== BigInt(0)) {
```

File: packages/vm/tests/api/EIPs/eip-3074-authcall.spec.ts (hunk 2 above)
Comment (ryanio) [reply packages/vm/tests/api/EIPs/eip-3074-authcall.spec.ts:475]: there are few more places like this, can help point them out or i can fix if you'd like.

another nit if you could use `st.equal` rather than `st.equals` we've normalized the repo on the former :)

File: packages/vm/src/evm/opcodes/gas.ts [hunk 4]
```diff
@@ -428,6 +428,58 @@ export const dynamicGasHandlers: Map<number, AsyncDynamicGasHandler> = new Map<
...
+        }
+      }
+
+      let gasLimit = maxCallGas(
+        runState.eei.getGasLeft() - gas,
+        runState.eei.getGasLeft() - gas,
+        runState,
+        common
+      )
+      if (!(currentGasLimit === BigInt(0))) {
```
Comment (ryanio) [inline packages/vm/src/evm/opcodes/gas.ts:472]: nit
```suggestion
      if (currentGasLimit !== BigInt(0)) {
```

File: packages/vm/tests/api/EIPs/eip-3074-authcall.spec.ts [hunk 5]
```diff
@@ -0,0 +1,597 @@
...
+      to: contractAddress,
+      gasLimit: 1000000,
+      gasPrice: 10,
+    }).sign(callerPrivateKey)
+
+    const result = await vm.runTx({ tx, block })
+    st.ok(
+      result.execResult.exceptionError?.error === ERROR.AUTHCALL_UNSET,
+      'threw with right error'
+    )
```
Comment (ryanio) [inline packages/vm/tests/api/EIPs/eip-3074-authcall.spec.ts:509]: can also use st.equal with string error msg comparison

```suggestion
    st.equal(result.execResult.exceptionError?.error, ERROR.AUTHCALL_UNSET, 'threw with right error')
```

**Comment by ryanio** [comment]
super cool, clean implementation as always!

**Comment by jochem-brouwer** [comment]
Cool, I have written the tests which I wanted, so this is now ready for review. Please let me know if you need help reviewing those bytecode tests.

File: packages/vm/src/evm/opcodes/EIP2929.ts [hunk 6]
```diff
@@ -18,7 +18,7 @@ export function accessAddressEIP2929(
   address: Address,
   common: Common,
   chargeGas = true,
-  isSelfdestruct = false
+  isSelfdestructOrAuthCall = false
```
Comment (ryanio) [inline packages/vm/src/evm/opcodes/EIP2929.ts:21]: should this be `isSelfdestructOrAuthcall`? we should ensure we are using the same case style for every use.

File: packages/vm/tests/api/EIPs/eip-3074-authcall.spec.ts [hunk 7]
```diff
@@ -0,0 +1,680 @@
...
+  const message = Buffer.concat([Buffer.from('03', 'hex'), chainId, paddedInvokerAddress, commit])
+  const msgHash = keccak256(message)
+  return ecsign(msgHash, privateKey, 0)
+}
+
+/**
+ * This method returns the bytecode in order to set AUTH
+ * @param commitUnpadded - The commit
+ * @param signature - The signature as obtained by `signMessage`
+ * @returns
```
Comment (ryanio) [inline packages/vm/tests/api/EIPs/eip-3074-authcall.spec.ts:75]: empty `@returns` here (and L110), you can just remove and typedoc knows it returns a Buffer, unless you want to add more description (same with the params). i do know this is in a test file though so just a nit :)

File: packages/vm/tests/api/EIPs/eip-3074-authcall.spec.ts [hunk 8]
```diff
@@ -0,0 +1,680 @@
...
+// Bytecode to store CALLER in slot 0 and GAS in slot 1 and the first 32 bytes of the input in slot 2
+// Returns the entire input as output
+const STORECALLER = Buffer.from('5A60015533600055600035600255366000600037366000F3', 'hex')
+
+/**
+ * This signs a message to be used for AUTH opcodes
+ * @param commitUnpadded - The commit which we want to sign
+ * @param address - The contract address we are using AUTH on
+ * @param privateKey - The private key of the account to sign
+ * @returns - Returns the signed message
```
Comment (ryanio) [inline packages/vm/tests/api/EIPs/eip-3074-authcall.spec.ts:60]: nit, no need for extra dash (same with all the params, but it doesn't seem VS Code adds an extra `-` like in this case)

> Screen Shot 2022-03-24 at 12 19 26 PM

```suggestion
 * @returns The signed message
```

File: packages/vm/tests/api/EIPs/eip-3074-authcall.spec.ts [hunk 9]
```diff
@@ -0,0 +1,680 @@
...
+})
+
+// Setups the environment for the VM, puts `code` at contractAddress and also puts the STORECALLER bytecode at the contractStorageAddress
+async function setupVM(code: Buffer) {
+  const vm = new VM({ common })
+  await vm.stateManager.putContractCode(contractAddress, code)
+  await vm.stateManager.putContractCode(contractStorageAddress, STORECALLER)
+  const account = await vm.stateManager.getAccount(callerAddress)
+  account.balance = PREBALANCE
+  await vm.stateManager.putAccount(callerAddress, account)
```
Comment (ryanio) [inline packages/vm/tests/api/EIPs/eip-3074-authcall.spec.ts:286]: instead of this pattern you can use modifyAccountFields now:

```suggestion
  await vm.stateManager.modifyAccountFields(callerAddress, { balance: PREBALANCE })
```

**Comment by jochem-brouwer** [comment]
Addressed your items @ryanio :smile: 

Seems like I have a lot of merge conflicts now. How do I fix this? Should I cherry-pick all commits on top of the new develop? Otherwise I think I am gonna do that rebase work twice (this should already be in `develop` ?)

**Comment by ryanio** [comment]
🙏 yes cherry picking on new develop seems to be easiest way, i just did it on my branch and had no conflicts

**Review by ryanio — APPROVED** [review]
LGTM! implementation, tests were all really easy to follow and comprehensive. re-checked the spec and everything seems covered. super exciting!

**Comment by holgerd77** [comment]
Rebased this via UI (did a test rebase locally).

**Comment by holgerd77** [comment]
Ok, given the existing approval I would now merge this in. Also had a very very rough look - without going into *any* implementation details, can agree, this looks extremely clean, with a substantial amount of tests added.

Also think that this PR shouldn't cause too much problems from a rebase perspective since functionality is added extremely locally/grouped.

So, thanks a lot @jochem-brouwer, great work on this! 😄 🎉

**Comment by cyrusadkisson** [comment]
Hi. I'm an app/solidity dev but have an interest in 3074 implementations due to a technical quirk of a contract I deployed in 2015. 

I'm keen to see the following in 3074 evm client implementations:
https://ethereum-magicians.org/t/eip-3074-auth-and-authcall-opcodes/4880/103

TL;DR - I think it makes the most sense, in an AUTHCALL, for tx.origin to be set to the address of the account being remote-controlled, not the account initializing the transaction (i.e. paying the gas) as it is, in spirit, the remote-controlled address performing the action.

I'm way out of my depth with your code, so a question: how does your implementation handle tx.origin?

**Comment by jochem-brouwer** [comment]
The implementation will follow the spec as it is currently written in the EIP, i.e. `tx.origin` is the account which has instantiated the original transaction, i.e. also the account paying for gas.

**Comment by cyrusadkisson** [comment]
Ok so I need to get it changed in the EIP first. Thanks.

Do you have an opinion either way? Seems to me that if tx.origin were meant to strictly be the gas payer, it would have been called "tx.gaspayer" instead.

**Comment by jochem-brouwer** [comment]
My initial opinion would be that we should not change the semantics of the origin opcode, I think it is generally agreed that this is the transaction origin thus the gaspayer. If you want the actual remote-controlled address (i.e. the AUTH address) then in the initial CALL layer you can use the `CALLER` opcode. If you also want this in deeper layers then you should probably add another opcode. This however begs the question, what should this opcode do in the situation if you first AUTHCALL from account A and then in this layer you AUTHCALL from account B. What should the opcode return, and why?

I think this is a rather complex situation which I don't have a strong opinion on yet. What motivates you to introduce these rules?

**Comment by cyrusadkisson** [comment]
Well this is the issue: Before 3074 became a thing, the transaction originator and gaspayer were one and the same. 3074 seeks to separate these identities. Some of the contracts I wrote in 2015 used tx.origin under the assumption of the former (transaction originator... it is, after all called "tx.origin" amirite?) and depend on this logic. If 3074 enables tx.origin/gaspayer to be someone other than the account originating the action, that breaks my contracts. 

I'm open to two solutions:
1. Make tx.origin the account from which the action originates. To me, this is logical anyway. (maybe a new tx.gaspayer opcode is created to cement the distinction)
2. If I'm the only party affected by this and everyone wants tx.origin to be the gaspayer, then we could do an irregular state change of /s/tx.origin/msg.sender on 2 of my 2015 contracts.
